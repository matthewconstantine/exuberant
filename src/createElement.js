import Root from './HostComponents/root'
import Project from './HostComponents/project'
import Dir from './HostComponents/dir'
import File from './HostComponents/file'
import Copy from './HostComponents/copy'

/**
 * Creates a HostComponent instance
 * @param {string} type Element type
 * @param {Object} props Component props
 * @param {Object} root Root instance
 */
function createElement(type, props, root) {
  console.log("type, props", type, props)
  
  const COMPONENTS = {
    ROOT: () => new Root(root, props),
    project: () => new Project(root, props),
    dir: () => new Dir(root, props),
    file: () => new File(root, props),
    copy: () => new Copy(root, props),
    default: undefined,
  }

  return COMPONENTS[type]() || COMPONENTS.default
}

export default createElement
