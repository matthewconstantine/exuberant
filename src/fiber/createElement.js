import Root from './HostComponents/Root'
import Dir from './HostComponents/Dir'
import File from './HostComponents/File'

/**
 * Creates a HostComponent instance
 * @param {string} type Element type
 * @param {Object} props Component props
 * @param {Object} root Root instance
 */
function createElement(type, props, root) {
  // console.log('CreateElement: ', {type, props, root})

  const COMPONENTS = {
    ROOT: () => new Root(root, props),
    DIR: () => new Dir(root, props),
    FILE: () => new File(root, props),
    default: undefined,
  }

  return COMPONENTS[type]() || COMPONENTS.default
}

export default createElement
