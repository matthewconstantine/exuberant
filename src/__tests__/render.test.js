import React from 'react'
import fs from 'fs-extra'
import { renderElement, rerenderElement } from '../'
import createElement from '../createElement'

const renderChange = (oldElement, newElement) =>
  rerenderElement(newElement, renderElement(oldElement, 'output'))

const Text = ({ children }) => (children.join ? children.join('\n') : children)

describe('Render', () => {
  beforeEach(() => fs.resetMock())

  describe('<root>', () => {
    it('Throws when no path is provided', () => {
      const element = <root />
      const shouldError = () => {
        createElement(element, 'output')
      }
      expect(shouldError).toThrowErrorMatchingSnapshot()
    })

    it('Renames children', () => {
      const before = <file name="Sportello" />
      const after = <file name="Doc" />
      renderChange(before, after)
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Removes children', () => {
      const before = <file name="Sportello" />
      const after = null
      renderChange(before, after)
      expect(fs.snapshot()).toMatchSnapshot()
    })
  })

  describe('<project>', () => {
    it('Creates itself', () => {
      const element = <project />
      renderElement(element, 'output')
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Adds children', () => {
      const before = (
        <project>
          <file name="Sportello" />
        </project>
      )
      const after = (
        <project>
          <file name="Sportello" />
          <file name="Shasta" />
        </project>
      )
      renderChange(before, after)
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Renames children', () => {
      const before = (
        <project>
          <file name="Sportello" />
        </project>
      )
      const after = (
        <project>
          <file name="Doc" />
        </project>
      )
      renderChange(before, after)
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Removes children', () => {
      const before = (
        <project>
          <file name="Sportello" />
        </project>
      )
      const after = <project />
      renderChange(before, after)
      expect(fs.snapshot()).toMatchSnapshot()
    })
  })

  describe('<dir>', () => {
    it('Creates itself', () => {
      const element = <dir name="shasta" />
      renderElement(element, 'output')
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Renames itself', () => {
      const first = <dir name="shasta" />
      const second = <dir name="faye" />
      renderChange(first, second)
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Deletes itself', () => {
      const first = (
        <dir name="wrapper">
          <dir name="deleteMe" />
        </dir>
      )
      const second = <dir name="wrapper" />
      renderChange(first, second)
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Throws when no name is provided', () => {
      const element = <dir />
      const shouldError = () => {
        renderElement(element, 'output')
      }
      expect(shouldError).toThrowErrorMatchingSnapshot()
    })

    it('Adds children', () => {
      const before = (
        <dir name="channel-estates">
          <file name="Sportello" />
        </dir>
      )
      const after = (
        <dir name="channel-estates">
          <file name="Sportello" />
          <file name="Shasta" />
        </dir>
      )
      renderChange(before, after)
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Renames children', () => {
      const before = (
        <dir name="channel-estates">
          <file name="Sportello" />
        </dir>
      )
      const after = (
        <dir name="channel-estates">
          <file name="Doc" />
        </dir>
      )
      renderChange(before, after)
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Removes children', () => {
      const before = (
        <dir name="channel-estates">
          <file name="Sportello" />
        </dir>
      )
      const after = <dir name="channel-estates" />
      renderChange(before, after)
      expect(fs.snapshot()).toMatchSnapshot()
    })
  })


  describe('<file>', () => {
    it('Creates a file', () => {
      const element = <file name="Sportello" />
      renderElement(element, 'output')
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Renames a file', () => {
      const first = <file name="Doc" />
      const second = <file name="Sportello" />
      renderChange(first, second)
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Renames a file and retains the correct name', () => {
      const first = <file name="Doc" />
      const second = <file name="Sportello" />
      renderChange(first, second)
      renderElement(second, 'output')
      renderElement(second, 'output')
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Deletes a file', () => {
      const first = (
        <dir name="wrapper">
          <file name="deleteMe" />
        </dir>
      )
      const second = <dir name="wrapper" />
      renderChange(first, second)
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Throws when no name is provided', () => {
      const element = <file />
      const shouldError = () => {
        renderElement(element, 'output')
      }
      expect(shouldError).toThrowErrorMatchingSnapshot()
    })

    it('Renders plain text', () => {
      const element = <file name="Sportello">Ahhh!</file>
      renderElement(element, 'output')
      expect(fs.getPath('output/Sportello')).toMatchSnapshot()
    })

    it('Mutates plain text', () => {
      const first = <file name="Sportello">Ahhh!</file>
      const second = <file name="Sportello">Whats that?</file>
      renderChange(first, second)
      expect(fs.getPath('output/Sportello')).toMatchSnapshot()
    })

    it('Deletes plain text', () => {
      const first = <file name="Sportello">Ahhh!</file>
      const second = <file name="Sportello" />
      renderChange(first, second)
      expect(fs.getPath('output/Sportello')).toMatchSnapshot()
    })

    it('Deletes multi-child text', () => {
      const first = (
        <file name="Sportello">
          <Text>Text</Text>
          <Text>deleteMe</Text>
        </file>
      )
      const second = (
        <file name="Sportello">
          <Text>Text</Text>
        </file>
      )
      renderChange(first, second)
      expect(fs.getPath('output/Sportello')).toMatchSnapshot()
    })

    it('Reorders multi-child text', () => {
      const first = (
        <file name="Sportello">
          <Text>First</Text>
          <Text>Second</Text>
          <Text>Third</Text>
        </file>
      )
      const second = (
        <file name="Sportello">
          <Text>First</Text>
          <Text>Third</Text>
          <Text>Second</Text>
        </file>
      )
      renderChange(first, second)
      expect(fs.getPath('output/Sportello')).toMatchSnapshot()
    })
  })
})
