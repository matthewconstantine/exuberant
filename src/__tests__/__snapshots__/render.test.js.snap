// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render <dir> Adds children 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "ensureDirSync",
      "path": "output/channel-estates",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
  ],
  "volume": Object {
    "output": true,
    "output/channel-estates": true,
    "output/channel-estates/Shasta": "",
    "output/channel-estates/Sportello": "",
  },
}
`;

exports[`Render <dir> Creates itself 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "ensureDirSync",
      "path": "output/shasta",
    },
  ],
  "volume": Object {
    "output": true,
    "output/shasta": true,
  },
}
`;

exports[`Render <dir> Deletes itself 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "ensureDirSync",
      "path": "output/wrapper",
    },
    Object {
      "action": "ensureDirSync",
      "path": "output/wrapper/deleteMe",
    },
    Object {
      "action": "removeSync",
      "path": "output/wrapper/deleteMe",
    },
  ],
  "volume": Object {
    "output": true,
    "output/wrapper": true,
  },
}
`;

exports[`Render <dir> Removes children 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "ensureDirSync",
      "path": "output/channel-estates",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "removeSync",
      "path": "output/channel-estates/Sportello",
    },
  ],
  "volume": Object {
    "output": true,
    "output/channel-estates": true,
  },
}
`;

exports[`Render <dir> Renames children 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "ensureDirSync",
      "path": "output/channel-estates",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "renameSync",
      "newPath": "output/channel-estates/Doc",
      "oldPath": "output/channel-estates/Sportello",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
  ],
  "volume": Object {
    "output": true,
    "output/channel-estates": true,
    "output/channel-estates/Doc": "",
  },
}
`;

exports[`Render <dir> Renames itself 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "ensureDirSync",
      "path": "output/shasta",
    },
    Object {
      "action": "renameSync",
      "newPath": "output/faye",
      "oldPath": "output/shasta",
    },
  ],
  "volume": Object {
    "output": true,
    "output/faye": true,
  },
}
`;

exports[`Render <dir> Throws when no name is provided 1`] = `"Dir does not have props.name"`;

exports[`Render <file> Creates a file 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
  ],
  "volume": Object {
    "output": true,
    "output/Sportello": "",
  },
}
`;

exports[`Render <file> Deletes a file 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "ensureDirSync",
      "path": "output/wrapper",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "removeSync",
      "path": "output/wrapper/deleteMe",
    },
  ],
  "volume": Object {
    "output": true,
    "output/wrapper": true,
  },
}
`;

exports[`Render <file> Deletes multi-child text 1`] = `"Text"`;

exports[`Render <file> Deletes plain text 1`] = `""`;

exports[`Render <file> Mutates plain text 1`] = `"Whats that?"`;

exports[`Render <file> Renames a file 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "renameSync",
      "newPath": "output/Sportello",
      "oldPath": "output/Doc",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
  ],
  "volume": Object {
    "output": true,
    "output/Sportello": "",
  },
}
`;

exports[`Render <file> Renames a file and retains the correct name 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "renameSync",
      "newPath": "output/Sportello",
      "oldPath": "output/Doc",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
  ],
  "volume": Object {
    "output": true,
    "output/Sportello": "",
  },
}
`;

exports[`Render <file> Renders plain text 1`] = `"Ahhh!"`;

exports[`Render <file> Reorders multi-child text 1`] = `
"First
Third
Second"
`;

exports[`Render <file> Throws when no name is provided 1`] = `"File node was not provided props.name"`;

exports[`Render <project> Adds children 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
  ],
  "volume": Object {
    "output": true,
    "output/Shasta": "",
    "output/Sportello": "",
  },
}
`;

exports[`Render <project> Creates itself 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
  ],
  "volume": Object {
    "output": true,
  },
}
`;

exports[`Render <project> Removes children 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "removeSync",
      "path": "output/Sportello",
    },
  ],
  "volume": Object {
    "output": true,
  },
}
`;

exports[`Render <project> Renames children 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "renameSync",
      "newPath": "output/Doc",
      "oldPath": "output/Sportello",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
  ],
  "volume": Object {
    "output": true,
    "output/Doc": "",
  },
}
`;

exports[`Render <root> Removes children 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "removeSync",
      "path": "output/Sportello",
    },
  ],
  "volume": Object {
    "output": true,
  },
}
`;

exports[`Render <root> Renames children 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "renameSync",
      "newPath": "output/Doc",
      "oldPath": "output/Sportello",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
  ],
  "volume": Object {
    "output": true,
    "output/Doc": "",
  },
}
`;

exports[`Render <root> Throws when no path is provided 1`] = `"COMPONENTS[type] is not a function"`;
